File: cmd/create/branch.go
package create

import (
	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/git"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func newBranch() *cobra.Command {
	var opts branchOptions

	c := &cobra.Command{
		Use:   "branch NAME",
		Short: "Create a new branch",

		Args: cobra.ExactArgs(1),

		ValidArgsFunction: cmd.NoneCompletion,
	}

	return cmd.Build(c, &opts)
}

type branchOptions struct {
	name string
}

func (o *branchOptions) Complete(c *cobra.Command, args []string) error {
	o.name = args[0]
	return nil
}

func (o *branchOptions) Run(ctx *context.Context) error {
	repo, err := repoutils.MustGetCurrentRepo(ctx)
	if err != nil {
		return err
	}
	err = ctx.SetRepo(repo)
	if err != nil {
		return err
	}

	gitCmd := git.WithPath(ctx.GetRepoPath())
	err = gitCmd.Run("checkout", "-b", o.name)
	if err != nil {
		return err
	}

	err = term.Confirm("Do you want to push branch %q to remote", o.name)
	if err == nil {
		return gitCmd.Run("push", "--set-upstream", "origin", o.name)
	}

	return nil
}

File: src/cmd/create/branch.rs
use anyhow::Result;
use async_trait::async_trait;
use clap::Args;

use crate::cmd::Command;
use crate::config::context::ConfigContext;
use crate::exec::git::branch::Branch;
use crate::{debug, output};

/// Create a new branch and push it to the remote.
#[derive(Debug, Args)]
pub struct CreateBranchCommand {
    /// Name of the branch to create.
    pub branch: String,
}

#[async_trait]
impl Command for CreateBranchCommand {
    fn name() -> &'static str {
        "branch"
    }

    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Running create branch command: {:?}", self);

        let branches = Branch::list(ctx.git())?;
        for branch in branches {
            if branch.name == self.branch {
                output!("Branch {:?} already exists", self.branch);
                return Ok(());
            }
        }

        ctx.git().execute(
            ["checkout", "-b", &self.branch],
            format!("Creating branch {}", self.branch),
        )?;

        ctx.git().execute(
            ["push", "--set-upstream", "origin", &self.branch],
            format!("Pushing branch {} to remote", self.branch),
        )?;

        Ok(())
    }
}
